Challenge 5 migration to extension

damianmarti opened the pull request:
Challenge 5 migrated to an extension following https://github.com/scaffold-eth/se-2-challenges/issues/234

Install from create-eth repo with:

```
yarn cli -e scaffold-eth/se-2-challenges:ch5-extension-init
```

closes #246

rin-st commented:
Houston, we have a problem

https://github.com/scaffold-eth/se-2-challenges/blob/challenge-5-state-channels/packages/nextjs/services/web3/wagmiConnectors.tsx#L1

In this challenge we don't use `burner-wallet` package, I used sessionStorage version instead of localStorage used in `burner-wallet`. I did it because we needed to get a new user in every browser tab.

So, options we have
- (most easy) ask to open Rubes in a new private browser windows
- update `burner-wallet` and add possibility to use sessionStorage
- try to use the same sessionStorage version
- ... something different

rin-st commented on extension/README.md.args.mjs:
```diff
@@ -0,0 +1,275 @@
+
+## Checkpoint 0: 📦 Environment 📚
+
+> in the same terminal, start your local network (a blockchain emulator in your computer):
```
remove "In the same terminal" part

damianmarti replied:
Fixed!

Yes, it is the same as in some of the previous extensions I will fix it there too!

Pabl0cks replied:
Can you check if the update is pushed?

damianmarti replied:
Sorry, I don't know what happened. Pushed now.

damianmarti commented:
> Houston, we have a problem
> 
> https://github.com/scaffold-eth/se-2-challenges/blob/challenge-5-state-channels/packages/nextjs/services/web3/wagmiConnectors.tsx#L1
> 
> In this challenge we don't use `burner-wallet` package, I used sessionStorage version instead of localStorage used in `burner-wallet`. I did it because we needed to get a new user in every browser tab.
> 
> So, options we have
> 
> * (most easy) ask to open Rubes in a new private browser windows
> * update `burner-wallet` and add possibility to use sessionStorage
> * try to use the same sessionStorage version
> * ... something different

oh! 

I think that we can go with the easier solution, asking to open the other user in a private browser window...

rin-st commented:
> I think that we can go with the easier solution, asking to open the other user in a private browser window...

Yes, let's try to do it that way. Update Readme with that info please. Keep in mind we need a new private browser window for every rube, different tabs of one window also doesn't work

damianmarti commented:
> > I think that we can go with the easier solution, asking to open the other user in a private browser window...
> 
> Yes, let's try to do it that way. Update Readme with that info please. Keep in mind we need a new private browser window for every rube, different tabs of one window also doesn't work

Checking the README I read:

⚠️ Note: previous challenges created new addresses by opening an incognito window or a different browser. This will not work for this challenge, because the off-chain application uses a very simple communication pipe that doesn't work between different browsers or private windows.

What is the limitation to making it work on different browsers or private windows?

rin-st commented:
Oh, it uses https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API (grep `BroadcastChannel` in the app).

As I remember it worked only with tabs before, but now I see in the docs:

"The Broadcast Channel API allows basic communication between [browsing contexts](https://developer.mozilla.org/en-US/docs/Glossary/Browsing_context) (that is, **windows**, tabs, frames, or iframes) and workers on the same [origin](https://developer.mozilla.org/en-US/docs/Glossary/Origin)."

So probably windows also works. Will try it tomorrow

rin-st commented:
Tried fast check, doesn't work (. Will play with it tomorrow

Also, this challenge uses `ethers`, `humanize-duration` and `@types/humanize-duration` deps. Uncomment checkpoint 4 and you'll see

damianmarti commented:
> Tried fast check, doesn't work (. Will play with it tomorrow

I tried the broadcast channel with the incognito window and it's not working :-(

We will need to go with one of the other options:

1. update burner-wallet and add the possibility to use sessionStorage
2. try to use the same sessionStorage version
3. ... something different

Maybe we can try option 1. 

Another option is to remove the broadcast channel and use a backend for the messages. 

But adding the sessionStorage to the burner wallet may be something useful for other projects too.

@rin-st what do you think?

rin-st commented:
> Maybe we can try option 1.

Yes, I'll try to update it later today

damianmarti commented:
> > Maybe we can try option 1.
> 
> Yes, I'll try to update it later today

Great! Thanks!

I will try to remove the ethers dependency, it seems like the missing function from viem is already implemented https://viem.sh/docs/utilities/parseSignature

damianmarti commented:
Looking at the burner wallet code from the previous challenge 5

> > > Maybe we can try option 1.
> > 
> > 
> > Yes, I'll try to update it later today
> 
> Great! Thanks!
> 
> I will try to remove the ethers dependency, it seems like the missing function from viem is already implemented https://viem.sh/docs/utilities/parseSignature

Removed viem dependency using hexToSignature (we should update it to use parseSignature when updating viem).

Added humanize-duration dependency too and fixed some minor code comments.

rin-st commented:
> Removed viem dependency using hexToSignature (we should update it to use parseSignature when updating viem).

Great! But why not to use `parseSignature` right away? Our `viem` version allows it

We updated `burner-connector` to next version so it can work with sessionStorage. I tried it with this challenge and now it works as expected. But to make it installed by default we need to wait for https://github.com/scaffold-eth/scaffold-eth-2/pull/1021 to be merged. Also, after that merge, we need to add this to this challenge READMESs to make it work

---
add this line to `wagmiConnectors.ts` to use sessionStroage version and test: 

```ts
rainbowkitBurnerWallet.useSessionStorage = true;
```

damianmarti commented:
> > Removed viem dependency using hexToSignature (we should update it to use parseSignature when updating viem).
> 
> Great! But why not to use `parseSignature` right away? Our `viem` version allows it

I think the viem version we are using does not have parseSignature yet. I tried it before changing to hexToSignature and the method does not exist.

> 
> We updated `burner-connector` to next version so it can work with sessionStorage. I tried it with this challenge and now it works as expected. But to make it installed by default we need to wait for [scaffold-eth/scaffold-eth-2#1021](https://github.com/scaffold-eth/scaffold-eth-2/pull/1021) to be merged. 

Great!! Thanks!!

Also, after that merge, we need to add this to this challenge READMESs to make it work
> 
> add this line to `wagmiConnectors.ts` to use sessionStroage version and test:
> 
> ```ts
> rainbowkitBurnerWallet.useSessionStorage = true;
> ```

I will add this line to the READMEs!

damianmarti commented:
@rin-st Added the rainbowkitBurnerWallet config to READMEs!

The PR you mentioned was merged, but now we have to wait to get this update on create-eth, right?

rin-st commented:
> I think the viem version we are using does not have parseSignature yet. I tried it before changing to hexToSignature and the method does not exist.

It works for me :shrug: . Added it [bb5cdf9](https://github.com/scaffold-eth/se-2-challenges/pull/250/commits/bb5cdf985ac7dbebb51ac2284f9225cba5218ba7), please try on your side 

> The PR you mentioned was merged, but now we have to wait to get this update on create-eth, right?

Yes, sorry :)

damianmarti commented:
> > I think the viem version we are using does not have parseSignature yet. I tried it before changing to hexToSignature and the method does not exist.
> 
> It works for me 🤷 . Added it [bb5cdf9](https://github.com/scaffold-eth/se-2-challenges/pull/250/commits/bb5cdf985ac7dbebb51ac2284f9225cba5218ba7), please try on your side

Working fine! Thanks!!

damianmarti commented:
> 
> > The PR you mentioned was merged, but now we have to wait to get this update on create-eth, right?
> 
> Yes, sorry :)

Create-eth with the new burner connector was released and working great with the challenge! cc/ @rin-st @Pabl0cks

Pabl0cks commented on extension/packages/nextjs/app/streamer/page.tsx:
```diff
@@ -0,0 +1,97 @@
+
+  useEffect(() => {
+    if (closedHistoryData?.length !== undefined && closedHistoryData?.length !== closed.length) {
+      const closedChannelsAddresses = closedHistoryData?.map(event => event.args[0]);
```
Same:

```
const closedChannelsAddresses = closedHistoryData?.map(event => event.args[0]) as AddressType[];
```

damianmarti replied:
Added type.

Pabl0cks commented on extension/packages/nextjs/app/streamer/page.tsx:
```diff
@@ -0,0 +1,97 @@
+
+  useEffect(() => {
+    if (openedHistoryData?.length !== undefined && openedHistoryData?.length !== opened.length) {
+      const openedChannelsAddresses = openedHistoryData?.map(event => event.args[0]).reverse();
```
Same:

```
const openedChannelsAddresses = openedHistoryData?.map(event => event.args[0]).reverse() as AddressType[];
```

damianmarti replied:
Added type.

Pabl0cks reviewed (commented):
Nice job @damianmarti and @rin-st !! 

With latest create-eth update and the new `rainbowkitBurnerWallet.useSessionStorage = true;` Readme hint is working great to me. 

Only added a couple nitpick comments.

rin-st reviewed (approved):
Lgtm! Just update types Pablo mentioned, please 🙂

damianmarti commented on extension/packages/nextjs/app/streamer/page.tsx:
```diff
@@ -0,0 +1,97 @@
+
+  useEffect(() => {
+    if (challengedHistoryData?.length !== undefined && challengedHistoryData.length !== challenged.length) {
+      const challengedChannelsAddresses = challengedHistoryData?.map(event => event.args[0]);
```
Did you get these errors when deploying to Vercel? I wasn't getting these errors, but I added the type anyway, it makes sense.

Thanks!!

Pabl0cks replied:
Yep! Vercel was failing to me

damianmarti commented:
> Nice job @damianmarti and @rin-st !!
> 
> With latest create-eth update and the new `rainbowkitBurnerWallet.useSessionStorage = true;` Readme hint is working great to me.
> 
> Only added a couple nitpick comments.

Thanks!!

Fixed! Please take a final review when you have some time. Thanks!

Pabl0cks reviewed (approved):
LGTM! Thx Damu 🙏
